### PR TITLE
Add `AuthorityHistoryPairFactory`

### LINKED ISSUE #71: Add `AuthorityWithHistory` factory
## 📚 Context

We envision to have factories for `Authority` and `History` separately, to leverage the optimal modularity of on-chain components. However, in order for a validator to deploy an Authority contract and a History contract and link them together, they would need to submit three separate transactions. This can be very pricey, depending on the network in which the contracts are deployed.

## ✔️ Solution

Create a smart contract that allows one to perform all three operations mentioned above in a single transaction. We will call this contract `AuthorityWithHistoryFactory`. This contract should also provide entry points for the deterministic deployment of the individual contracts.

## ⛓ Requires

- [ ] #59
- [ ] #70

## 📈 Subtasks

- [ ] Define the interface
- [ ] Implement the interface
- [ ] Test the contract
- [ ] Create deployment script

### PR DESCRIPTION
**Pull request by guidanoli**
Adds interface, implementation and tests.

Closes cartesi/rollups-contracts#29.

### CODE REVIEW

**Review by pedroargento — APPROVED** [review]

File: onchain/rollups/contracts/consensus/authority/AuthorityHistoryPairFactory.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+// specific language governing permissions and limitations under the License.
+
+pragma solidity ^0.8.8;
+
+import {IAuthorityHistoryPairFactory} from "./IAuthorityHistoryPairFactory.sol";
+import {Authority} from "./Authority.sol";
+import {IAuthorityFactory} from "./IAuthorityFactory.sol";
+import {History} from "../../history/History.sol";
+import {IHistoryFactory} from "../../history/IHistoryFactory.sol";
+import {Authority} from "./Authority.sol";
```
Comment (ZzzzHui) [inline onchain/rollups/contracts/consensus/authority/AuthorityHistoryPairFactory.sol:20]: this line is redundant

File: onchain/rollups/contracts/consensus/authority/IAuthorityHistoryPairFactory.sol [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    /// @notice Get the factory used to deploy `History` contracts
+    /// @return The `History` factory
+    function getHistoryFactory() external view returns (IHistoryFactory);
+
+    /// @notice Deploy a new authority-history pair.
+    /// @param _authorityOwner The initial authority owner
+    /// @param _inputBox The input box
+    /// @return The authority
+    /// @return The history
+    /// @dev On success, MUST emit an `AuthorityHistoryPairCreated` event.
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/authority/IAuthorityHistoryPairFactory.sol:49]: Oops, you're right!

File: onchain/rollups/contracts/consensus/authority/IAuthorityHistoryPairFactory.sol [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        IInputBox _inputBox
+    ) external returns (Authority, History);
+
+    /// @notice Deploy a new authority-history pair deterministically.
+    /// @param _authorityOwner The initial authority owner
+    /// @param _inputBox The input box
+    /// @param _salt The salt used to deterministically generate the authority-history pair address
+    /// @return The authority
+    /// @return The history
+    /// @dev On success, MUST emit an `AuthorityHistoryPairCreated` event.
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/authority/IAuthorityHistoryPairFactory.sol:61]: I will also remove this line.

File: onchain/rollups/contracts/consensus/authority/AuthorityHistoryPairFactory.sol (hunk 1 above)
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/authority/AuthorityHistoryPairFactory.sol:20]: Oops, you're right!

**Review by ZzzzHui — APPROVED** [review]

**Review by DaniOrtegaB — COMMENTED** [review]
Great work and excellent code.

File: onchain/rollups/test/foundry/consensus/authority/AuthorityHistoryPairFactory.t.sol [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+import {AuthorityHistoryPairFactory} from "contracts/consensus/authority/AuthorityHistoryPairFactory.sol";
+import {IAuthorityFactory} from "contracts/consensus/authority/IAuthorityFactory.sol";
+import {AuthorityFactory} from "contracts/consensus/authority/AuthorityFactory.sol";
+import {Authority} from "contracts/consensus/authority/Authority.sol";
+import {IHistoryFactory} from "contracts/history/IHistoryFactory.sol";
+import {HistoryFactory} from "contracts/history/HistoryFactory.sol";
+import {History} from "contracts/history/History.sol";
+import {IInputBox} from "contracts/inputs/IInputBox.sol";
+import {InputBox} from "contracts/inputs/InputBox.sol";
+import {Vm} from "forge-std/Vm.sol";
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/test/foundry/consensus/authority/AuthorityHistoryPairFactory.t.sol:26]: Makes sense. Thanks!

**Comment by guidanoli** [comment]
Rebased.

**Comment by guidanoli** [comment]
Rebased.

**Review by DaniOrtegaB — APPROVED** [review]

**Comment by guidanoli** [comment]
This PR has been migrated to https://github.com/cartesi/rollups-contracts/pull/30.

**Pull request closed without merging**